ui/src/dashboards/components/GraphOptionsCustomizableField.tsx

```diff
@@ -1,70 +1,186 @@
 import React, {PureComponent} from 'react'
+import {findDOMNode} from 'react-dom'
+import {
+  DragSourceSpec,
+  DropTargetConnector,
+  DragSourceMonitor,
+  DragSource,
+  DropTarget,
+  DragSourceConnector,
+} from 'react-dnd'
 
-import InputClickToEdit from 'src/shared/components/InputClickToEdit'
+const FieldType = 'field'
 
 interface Field {
   internalName: string
   displayName: string
   visible: boolean
+  order?: number
+  index?: number
 }
 
-interface Props {
+interface FieldSourceProps {
   internalName: string
   displayName: string
   visible: boolean
-  onFieldUpdate: (field: Field) => void
+  index: number
+  id: string
+  key: string
+  isDragging?: boolean
+  onFieldUpdate?: (field: Field) => void
+  connectDragSource?: any
+  connectDropTarget?: any
+  connectDragPreview?: any
+  moveField: (dragIndex: number, hoverIndex: number) => void
 }
 
-class GraphOptionsCustomizableField extends PureComponent<Props, {}> {
+const fieldSource: DragSourceSpec<FieldSourceProps> = {
+  beginDrag(props) {
+    return {
+      id: props.id,
+      index: props.index,
+    }
+  },
+}
+
+const fieldTarget = {
+  hover(props, monitor, component) {
+    const dragIndex = monitor.getItem().index
+    const hoverIndex = props.index
+
+    // Don't replace items with themselves
+    if (dragIndex === hoverIndex) {
+      return
+    }
+
+    // Determine rectangle on screen
+    const hoverBoundingRect = findDOMNode(component).getBoundingClientRect()
```

Comment:
is the the rectangle around the current component or the target location of a component? could you make the comment or variable name a bit more descriptive?